### ISSUE TITLE
Split pointers for diagonal and off-diagonal parts of MatrixCSR

### ISSUE THREAD

**Issue by chrisrichardson** [FEniCS/dolfinx#2628]
For better compatibility with 3rd party libraries, it would be better to have the diagonal and off-diagonal blocks of the sparse matrix in `MatrixCSR` stored separately. The data structures would be the same, just two instead of one, and the storage should not be any more.

**Comment by garth-wells** [comment]
@chrisrichardson has this been completed?

**Comment by chrisrichardson** [comment]
No. I am also not convinced it is always the right solution. Some solvers use a unified format, some split. I have got a working prototype (split format) so maybe we can work on having both available  via an option or subclassing.

**Comment by chrisrichardson** [comment]
https://github.com/FEniCS/dolfinx/tree/chris/matrixcsr-bs-experiment

**Comment by garth-wells** [comment]
Splitting the matrix would make wrapping sparse matrices locally using SciPy more difficult.

**Comment by chrisrichardson** [comment]
I agree. I think we should consider supporting both formats internally.

**Comment by jorgensd** [comment]
@chrisrichardson  is this now resolved with: https://github.com/FEniCS/dolfinx/blob/f6a4a5a3de45003a38d3b8d3c667a1d1cf14adf9/cpp/dolfinx/la/MatrixCSR.h#L343-L359 ?

**Comment by chrisrichardson** [comment]
I am revisiting this whilst looking into more MatrixCSR operations (SpGEMM and transpose). I think splitting into diagonal and off-diagonal blocks *is* the right approach. Scipy is not used greatly in parallel and anyway, can be accommodated with two operations (diag and offdiag) separately.
Having separation will allow e.g. transposition of local diag block whilst using non-blocking MPI on offdiag...
